### PR TITLE
Add AppImage build

### PR DESCRIPTION
**Pull request by ZhymabekRoman**
Thanks for the nheko!

I made changes to the CI building so that nheko would also compile as a single executable AppImage file. It took a lot of work but it seems to work, more tests are needed. I tested it on Ubuntu 20.04 and Debian 11 under chroot. I had to drop the docker appimagecrafters/appimage-builder image in favor of Ubunutu 22.04 because it was based on Ubuntu 18.04.

If there are any issues with AppImage in the future, ping me please

### CODE REVIEW

**Comment by deepbluev7** [comment]
Okay, so huge disclaimer ahead:

I really don't like AppImages. I spent a long time maintaining one and it has never been working reliably for me and caused us a lot of issues. However: A lot of people ask for them and since you spent a lot of effort on this AND promised to maintain it, I don't see any reason not to merge it. I just want to make it clear, that I will advertise them as less supported and will be depending on you to fix issues, when they crop up. I don't really want people to download it thinking "cool, just a single runnable download, no need for stuff like flatpak!" and have that reflecting badly on Nheko, so there will be a disclaimer on the releases about them being less supported. i would also ask you to add the allow_failure keyword to this job, so that it failing doesn't mark the pipeline as failed: https://docs.gitlab.com/ee/ci/yaml/#allow_failure

Now about the actual feedback. Currently the pipeline for this is failing, partially because of the bundled libevent. I don't see why you need that bundled. You should be able to disable the bundled openssl, libevent, libcurl, qtkeychain and spdlog at least. If you do that, does the pipeline pass?

**Comment by ZhymabekRoman** [comment]
> Now about the actual feedback. Currently the pipeline for this is failing, partially because of the bundled libevent.

Hmmm, that's weird, I successfully build AppImage with these configurations on my GitLab account. 

> I don't see why you need that bundled.

If not bundle it then nheko crashes with signal 6. I will try again now

**Comment by ZhymabekRoman** [comment]
Pipeline failed, again:
```
Incorrect OPENSSL_VERSION_NUMBER define in header:
  /usr/include/openssl/opensslv.h
```
Ubuntu 22.04 seems to only have OpenSSL version 3.x, not 1.x. We need to compile it or install it from third-party PPAs

**Comment by deepbluev7** [comment]
openssl3 should work fine too with Nheko, but that might not run on all systems?

**Comment by deepbluev7** [comment]
Can you download and verify, that the appimage works? After that I think we can merge it :3

**Comment by deepbluev7** [comment]
So the appimage is currently 130 MB big, maybe we can exclude a few more directories, like /usr/include and these programs from /bin:

```
-rw-r--r-- 1 nicolas nicolas     6963  4. Jul 11:20 c_rehash
-rw-r--r-- 1 nicolas nicolas   342144 21. Okt 16:52 cmark
-rw-r--r-- 1 nicolas nicolas    22528 21. Okt 16:52 dbus-cleanup-sockets
-rw-r--r-- 1 nicolas nicolas   239680 21. Okt 16:52 dbus-daemon
-rw-r--r-- 1 nicolas nicolas    34816 21. Okt 16:52 dbus-monitor
-rw-r--r-- 1 nicolas nicolas    22528 21. Okt 16:52 dbus-run-session
-rw-r--r-- 1 nicolas nicolas    34816 21. Okt 16:52 dbus-send
-rw-r--r-- 1 nicolas nicolas    22528 21. Okt 16:52 dbus-update-activation-environment
-rw-r--r-- 1 nicolas nicolas    22528 21. Okt 16:52 dbus-uuidgen
-rw-r--r-- 1 nicolas nicolas  1013760 21. Okt 16:52 openssl
-rw-r--r-- 1 nicolas nicolas      154 21. Okt 16:51 qt.conf
-rw-r--r-- 1 nicolas nicolas    30656 21. Okt 16:52 session-migration
```


Also some stuff from /usr/share like libwacom, upstart, zsh, some stuff from the runtime folder. Maybe also some libraries from /usr/lib/x86_64-linux-gnu.

**Comment by ZhymabekRoman** [comment]
> Can you download and verify, that the appimage works?

AppImage works, tested using Ubuntu 20.04

> So the appimage is currently 130 MB big, maybe we can exclude a few more directories

I was able to shrink the AppImage 3 times (to 50 mb) by removing GStream, also accelerated startup time to 5 times, but multimedia files refuse to play because QT Multimedia uses GStream as backend in UNIX (https://wiki.qt.io/Qt_5.13_Multimedia_Backends).

**Comment by deepbluev7** [comment]
Yeah, removing all of gstreamer is not feasible. Although we possibly could remove some formats. Storing 130MB per build is somewhat expensive for my server :D

File: AppImageBuilder.yml [hunk 1]
```diff
@@ -0,0 +1,136 @@
...
+    - sourceline: deb http://archive.ubuntu.com/ubuntu/ jammy-backports main restricted universe multiverse
+    - sourceline: deb http://archive.canonical.com/ubuntu jammy partner
+      key_url: http://keyserver.ubuntu.com/pks/lookup?op=get&search=0x871920d1991bc93c
+    include:
+    # idk, need there ca-certificates, but some systems may not have
+    - ca-certificates
+    - gstreamer1.0-libav
+    - gstreamer1.0-plugins-bad
+    - gstreamer1.0-plugins-base
+    - gstreamer1.0-plugins-good
```
Comment (deepbluev7) [inline AppImageBuilder.yml:31]: I think these can actually be split up, we only need some of the packages from them.

**Comment by deepbluev7** [comment]
Well, you do need the webrtc deps to be able to do calls. But you can trim out some of the other gstreamer deps by not depending on the meta packages, like Nheko doesn't need dvd support in gstreamer.

**Comment by orowith2os** [comment]
Making nheko an AppImage seems like it takes a lot of effort to work and make efficient, at that point why not just settle for flatpak builds (which are already available, work on more platforms, and don't rely on host dependencies)?

It seems like a lot of unneeded effort, for little benefit (or, if anything, *no* benefit, as an AppImage would be more storage-heavy than a flatpak).

**Comment by ZhymabekRoman** [comment]
@deepbluev7, I don't think we can compress AppImage any more. I also tried using dwarfs instead of squashfs, it did not work. But I think ~65 megabytes is a good result compared to 130 megabytes (https://github.com/Nheko-Reborn/nheko/releases/tag/v0.8.2). I think we can merge

**Comment by deepbluev7** [comment]
I think you removing the gstreamer stuff broke voice calls and media playback, but I haven't specifically tested it. I thought I had mentioned that, but I guess I will just test it later :3

**Comment by ZhymabekRoman** [comment]
> broke voice calls

AppImage compiles without VOIP support. 

> media playback

opus and mp3 files open correctly

**Comment by deepbluev7** [comment]
I guess I need to advertise this not supporting voip then, but I guess that is fine.

**Comment by deepbluev7** [comment]
Sadly currently the CI for PRs is broken, however, it seems like the job I manually started failed for some weird reason? https://nheko.im/nheko-reborn/nheko/-/jobs/24213

**Comment by ZhymabekRoman** [comment]
Done, I think I fixed the problem

**Comment by deepbluev7** [comment]
Thanks for the quick response, let's see if this passes and I can test it :3 

https://nheko.im/nheko-reborn/nheko/-/jobs/24234

**Comment by deepbluev7** [comment]
Seems to be missing this package, so it starts with a blank screen when merged into master: https://packages.ubuntu.com/jammy/qml-module-qtquick-particles2

**Comment by ZhymabekRoman** [comment]
Maybe you have D-Bus root information transfer set to nheko in your settings?

**Comment by deepbluev7** [comment]
It might just be that my dbus daemon died again, do you get notifications when running the appimage?

**Comment by ZhymabekRoman** [comment]
Did you get the whole nheko starting? Were you able to log in to your account?

**Comment by deepbluev7** [comment]
I had it starting, but the screen is black because the particles qml module is missing. Otherwise it seemed to work fine, it connected to my existing account on my  test profile.

**Comment by ZhymabekRoman** [comment]
Fine, let's try to add the missing package then.

**Comment by deepbluev7** [comment]
https://nheko.im/nheko-reborn/nheko/-/jobs/24243

**Comment by ZhymabekRoman** [comment]
Done

**Comment by deepbluev7** [comment]
Hm, I still get these errors...

```
[2023-01-05 04:28:00.714] [ui] [warning] Could not connect to D-Bus!
[2023-01-05 04:28:15.379] [qml] [warning] qrc:/qml/Root.qml:352:9: Type ChatPage unavailable
             ChatPage {
             ^ (qrc:/qml/Root.qml:352, )
[2023-01-05 04:28:15.379] [qml] [warning] qrc:/qml/ChatPage.qml:105:17: Type RoomList unavailable
                     RoomList {
                     ^ (qrc:/qml/ChatPage.qml:105, )
[2023-01-05 04:28:15.380] [qml] [warning] qrc:/qml/RoomList.qml:98:17: Type TimelineView unavailable
                     TimelineView {
                     ^ (qrc:/qml/RoomList.qml:98, )
[2023-01-05 04:28:15.380] [qml] [warning] qrc:/qml/TimelineView.qml:17:1: module "QtQuick.Particles" is not installed
     import QtQuick.Particles 2.15
     ^ (qrc:/qml/TimelineView.qml:17, )
```

File: AppImageBuilder.yml [hunk 2]
```diff
@@ -0,0 +1,113 @@
...
+    - libnss-mdns
+    - libpcre3
+    - qml-module-qt-labs-platform
+    - qml-module-qt-labs-settings
+    - qml-module-qtgraphicaleffects
+    - qml-module-qtmultimedia
+    - qml-module-qtquick2
+    - qml-module-qtquick-controls2
+    - qml-module-qtquick-layouts
+    - qml-module-qtquick-window2
```
Comment (deepbluev7) [inline AppImageBuilder.yml:50]: You forgot the particles module here

Comment (ZhymabekRoman) [reply AppImageBuilder.yml:50]: Ohhh, yeahh

File: AppImageBuilder.yml [hunk 3]
```diff
@@ -0,0 +1,113 @@
...
+    - libxrender1
+    - libxshmfence1
+    - libxv1
+    - libxxf86vm1
+    - libzstd1
+    exclude: []
+  files:
+    exclude:
+    - usr/share/man
+    - usr/share/doc
```
Comment (deepbluev7) [inline AppImageBuilder.yml:89]:
```suggestion
    - usr/share/doc
    - usr/bin/cmark
    - usr/bin/dbus*
    - usr/bin/openssl
    - usr/bin/c_rehash
    - usr/bin/session-migration
    - usr/include
    - usr/sbin
    - lib/systemd
```

**Comment by deepbluev7** [comment]
Seems to work mostly properly now, you think I can merge it? Do you want to test it some more? The styling is broken, but afaik that is just an AppImage limitation. And I don't think notifications work, but I can just document that.

File: .gitlab-ci.yml [hunk 4]
```diff
@@ -223,6 +223,69 @@ build-flatpak-arm64:
...
+        -DVOIP=OFF -DMAN=OFF
+        -DCMAKE_BUILD_TYPE=Release -DHUNTER_CONFIGURATION_TYPES=Release
+        -DCI_BUILD=ON -DFETCHCONTENT_QUIET=OFF
+        -DJSON_ImplicitConversions=OFF
+    - DESTDIR=`pwd`/AppDir ninja -C build install/local
+    - DESTDIR=`pwd`/AppDir ninja -C build _deps/cmark-build/src/install
+    - mkdir -p AppDir/usr/lib/x86_64-linux-gnu AppDir/lib/x86_64-linux-gnu
+    - appimage-builder --skip-test
+  after_script:
+    - bash ./.ci/upload-nightly-gitlab.sh nheko-latest-x86_64.AppImage
```
Comment (deepbluev7) [inline .gitlab-ci.yml:277]:
```suggestion
    - bash ./.ci/upload-nightly-gitlab.sh nheko-latest-x86_64.AppImage
  rules:
    - if : '$CI_PIPELINE_TRIGGERED == null'
```

**Comment by ZhymabekRoman** [comment]
I have notifications working fine as they should. I think we can merge. The basic functionality of nheko works fine, once accidentally caught segfault, but maybe it has nothing to do with AppImage

**Comment by deepbluev7** [comment]
Then thank you very much for working on it <sub>and I hope you continue to fix issues with it</sub>!

**Pull request merged**